docs/csharp/advanced-topics/performance/index.md

```diff
@@ -0,0 +1,94 @@
+
+:::code language="csharp" source="./snippets/ref-safety/EscapeScopes.cs" id="RefSafeToEscapeScopes":::
+
+> [!NOTE]
+> When the `in` modifier is applied to a parameter, that parameter can be returned by `ref readonly`, not `ref`.
+
+The compiler ensures that a reference can't escape its *ref safe to escape scope*. You can use `ref` parameters, `ref return` and `ref` local variables safely because the compiler detects if you've accidentally written code where a `ref` expression could be accessed when its storage isn't valid.
+
+## Safe to escape scope and ref structs
+
+`ref struct` types require more rules to ensure they can be used safely. A `ref struct` type may include `ref` fields. That requires the introduction of a *safe to escape scope*. For most types, the *safe to escape scope* is the entire method. In other words, a value that's not a `ref struct` can always be returned from a method.
+
+Informally, the *safe to escape scope* for a `ref struct` is the scope where all of its `ref` fields can be accessed. In other words, it's the intersection of the *ref safe to escape scopes* of all its `ref` fields. The following method returns a `ReadOnlySpan<char>` to a member field, so its *safe to escape scope* is the method:
+
+:::code language="csharp" source="./snippets/ref-safety/EscapeScopes.cs" id="SafeToEscapeScope":::
+
+In contrast, the following code emits an error because the `ref field` member of the `Span<int>` refers to the stack allocated array of integers. It can't escape the method:
+
+```csharp
+public Span<int> M()
+{
+    int length = 3;
+    Span<int> numbers = stackalloc int[length];
+    for (var i = 0; i < length; i++)
+    {
+        numbers[i] = i;
+    }
+    return numbers; // Error! numbers can't escape this method.
+}
+```
+
+## Unify memory types
+
+The introduction of <xref:System.Span%601?displayProperty=fullName> and <xref:System.Memory%601?displayProperty=fullName> provide a unified model for working with memory. <xref:System.ReadOnlySpan%601?displayProperty=fullName> and <xref:System.ReadOnlyMemory%601?displayProperty=fullName> provide readonly versions for accessing memory. They all provide an abstraction over a block of memory storing an array of similar elements. The difference is that `Span<T>` and `ReadOnlySpan<T>` are `ref struct` types whereas `Memory<T>` and `ReadOnlyMemory<T>` are `struct` types. Spans contain a `ref field`. Therefore instances of a span can't leave its *ref safe to escape* context. The implementation of `Memory<T>` and `ReadOnlyMemory<T>` remove this restriction. You use these types to directly access memory buffers.
```

Comment:
>Therefore instances of a span can't leave its *ref safe to escape* context. [](http://example.com/codeflow?start=601&length=75)

Instances of a span can't leave its _safe to escape_ context (which is the _ref safe to escape_ of the ref field within the span).

Reply:
I took this comment verbatim. I think that's fine here.

I'll see if I get questions on how we define the *safe to escape* scope for a `ref struct` with multiple `ref` fields. Intuitively, it would be the "smallest" scope of the multiple *ref safe to escape* scopes. Do we already have formal language for that?